lib/seahorn/BvOpSem2TrackingRawMemMgr.cc

```diff
@@ -277,119 +282,149 @@ TrackingRawMemManager::memsetMetaData(MetadataKind kind, PtrTy ptr,
   // make sure we can fit the supplied value in metadata memory slot
   assert(llvm::Log2_64(val) + 1 <= g_MetadataBitWidth &&
          "Metadata cannot fit!");
-  return MemValTy(
-      kind, memIn,
-      m_metadata_map[kind]->MemSet(ptr, m_ctx.alu().ui(val, g_MetadataBitWidth),
-                                   len, memIn.getMetadata(kind),
-                                   m_metadata_map[kind]->wordSzInBytes()));
+  auto index = static_cast<size_t>(kind) + 1;
+  auto r =
+      hana::make_range(hana::int_c<0>, TrackingMemoryTuple::GetTupleSize());
+  auto r_t = hana::to_tuple(r);
+  auto z_t = hana::zip(r_t, hana::keys(m_submgrs));
+  auto c = hana::transform(z_t, [&](auto pair) {
+    auto current = hana::at(pair, hana::size_c<0>);
+    auto key = hana::at(pair, hana::size_c<1>);
+    if (current == index) {
+      return hana::at_key(m_submgrs, key)
+          .MemSet(ptr, m_ctx.alu().ui(val, g_MetadataBitWidth), len,
+                  memIn.getElementVal(index),
+                  hana::at_key(m_submgrs, key).wordSzInBytes());
+    } else {
+      return memIn.getElementVal(current);
+    }
+  });
+  return MemValTy(c);
 }
 Expr TrackingRawMemManager::getMetaData(MetadataKind kind, PtrTy ptr,
                                         MemValTy memIn, unsigned int byteSz) {
   // TODO: expose a method in OpSemMemManager to loadAlignedWordFromMem
-  return m_metadata_map.at(kind)->loadIntFromMem(
-      ptr, memIn.getMetadata(kind), byteSz,
-      m_metadata_map.at(kind)->wordSizeInBytes());
+  auto index = static_cast<size_t>(kind) + 1;
+  auto r =
+      hana::make_range(hana::int_c<0>, TrackingMemoryTuple::GetTupleSize());
+  auto r_t = hana::to_tuple(r);
+  Expr ret;
+  hana::for_each(hana::zip(r_t, hana::keys(m_submgrs)), [&](auto pair) {
```

Comment:
I'm confused why we need to do a linear search here to retrieve a sub mem manager.